File: plugins/citation-gate/hooks/citation-gate.js
#!/usr/bin/env node
/**
 * Stop Hook: Citation verification gate (non-skill writing path)
 * Event: Stop
 * 解析 transcript 找本会话改过的 .tex/.bib/.md，有引用就跑 citation_gate 校验器；
 * HARD_FAIL → block 返工（最多 MAX_ROUNDS 轮防死锁），否则放行。
 */
const fs = require('fs');
const os = require('os');
const path = require('path');
const { spawnSync } = require('child_process');

const HOME = os.homedir();
// Plugin root = the directory that contains this hook's `hooks/` folder.
// The bundled `citation_gate` Python package lives at <PLUGIN_ROOT>/citation_gate,
// so PYTHONPATH must be PLUGIN_ROOT for `python3 -m citation_gate` to resolve it.
const PLUGIN_ROOT = process.env.CLAUDE_PLUGIN_ROOT || path.resolve(__dirname, '..');
const LIB_DIR = PLUGIN_ROOT;
// Per-user state (block-round counters) stays under the user's home, not the plugin.
const ROUNDS_DIR = path.join(HOME, '.claude', '.cache', 'citation_gate', 'rounds');
const MAX_ROUNDS = 3;
const EXTS = new Set(['.tex', '.bib', '.md']);

function collectCandidateFiles(transcriptPath) {
  const seen = [];
  const add = (fp) => {
    if (fp && EXTS.has(path.extname(fp).toLowerCase())
        && fs.existsSync(fp) && !seen.includes(fp)) seen.push(fp);
  };
  let lines = [];
  try {
    lines = fs.readFileSync(transcriptPath, 'utf8').split('\n').filter(Boolean);
  } catch { return []; }
  for (const line of lines) {
    let ev;
    try { ev = JSON.parse(line); } catch { continue; }
    const content = ev && ev.message && ev.message.content;
    if (!Array.isArray(content)) continue;
    for (const block of content) {
      if (block && block.type === 'tool_use'
          && ['Write', 'Edit', 'MultiEdit'].includes(block.name)) {
        add(block.input && block.input.file_path);
      }
    }
  }
  return seen;
}

function hasCitationSignal(filePath) {
  let text = '';
  try { text = fs.readFileSync(filePath, 'utf8'); } catch { return false; }
  if (text.includes('<!-- citation-gate: skip -->')) return false;
  return /\\bibitem|\\cite|\[\d+\]|@(inproceedings|article|misc|book)\b/m.test(text);
}

function readRounds(sessionId) {
  try {
    return JSON.parse(fs.readFileSync(path.join(ROUNDS_DIR, `${sessionId}.json`), 'utf8')).n || 0;
  } catch { return 0; }
}
function bumpRounds(sessionId) {
  fs.mkdirSync(ROUNDS_DIR, { recursive: true });
  const n = readRounds(sessionId) + 1;
  fs.writeFileSync(path.join(ROUNDS_DIR, `${sessionId}.json`), JSON.stringify({ n }));
  return n;
}
function resetRounds(sessionId) {
  try { fs.unlinkSync(path.join(ROUNDS_DIR, `${sessionId}.json`)); } catch { /* noop */ }
}

function runVerifier(files) {
  const result = spawnSync('python3', ['-m', 'citation_gate', '--json', ...files], {
    env: { ...process.env, PYTHONPATH: LIB_DIR },
    encoding: 'utf8', timeout: 120000,
  });
  if (result.error) throw result.error;
  const out = result.stdout || '';
  if (!out.trim()) {
    throw new Error(`verifier produced no output (exit=${result.status}): ${(result.stderr || '').slice(0, 200)}`);
  }
  return JSON.parse(out);
}

function decide(report, roundCount, maxRounds) {
  const hard = report.hardFail || report.hard_fail || [];
  const soft = report.softWarn || report.soft_warn || [];
  if (hard.length === 0) {
    let msg = '引用校验通过。';
    if (soft.length) msg += ` ${soft.length} 条未能核实，已标 [unverified]，请人工确认。`;
    return { block: false, message: msg };
  }
  if (roundCount >= maxRounds) {
    return { block: false, message:
      `引用校验仍有 ${hard.length} 条疑似编造，但已达返工上限（${maxRounds} 轮），放行。请人工(manual)复核：\n`
      + hard.map((h) => `  [${h.index}] ${h.message}`).join('\n') };
  }
  const detail = hard.map((h) => `  [${h.index}] ${h.message}`).join('\n');
  return { block: true, message:
    `检测到 ${hard.length} 条疑似编造的引用，请逐条按权威记录修正后再交付：\n${detail}` };
}

function main() {
  let input = {};
  try {
    const raw = fs.readFileSync(0, 'utf8');
    if (raw.trim()) input = JSON.parse(raw);
  } catch { /* empty */ }

  const pass = (sysMsg) => {
    console.log(JSON.stringify(sysMsg ? { continue: true, systemMessage: sysMsg }
                                       : { continue: true }));
    process.exit(0);
  };

  if (process.env.CITATION_GATE === 'off') return pass();
  const sessionId = input.session_id || 'unknown';

  const files = collectCandidateFiles(input.transcript_path || '')
    .filter(hasCitationSignal);
  if (files.length === 0) { resetRounds(sessionId); return pass(); }

  let report;
  try {
    report = runVerifier(files);
  } catch (e) {
    return pass(`引用校验器未能运行（${(e && e.message || '').slice(0, 120)}），本次跳过。`);
  }

  const round = readRounds(sessionId);
  const d = decide(report, round, MAX_ROUNDS);
  if (d.block) {
    bumpRounds(sessionId);
    console.log(JSON.stringify({ decision: 'block', reason: d.message }));
    process.exit(0);
  }
  resetRounds(sessionId);
  return pass(d.message);
}

if (require.main === module) main();

module.exports = { collectCandidateFiles, hasCitationSignal, decide,
                   readRounds, bumpRounds, resetRounds };

File: plugins/citation-gate/hooks/citation-gate.py
#!/usr/bin/env python3
"""Stop hook (Python entry): citation verification gate for the non-skill writing path.

Replaces the previous Node.js hook so the plugin runs in runtimes that ship
``python3`` but not ``node`` (e.g. the AutoReproBackend agent runtime, whose
container has no node on PATH → the old `node …citation-gate.js` exited 127).

解析 transcript 找本会话改过的 .tex/.bib/.md,有引用就跑 bundled ``citation_gate``
校验器(**进程内 import**,不再 subprocess/依赖 node);HARD_FAIL → block 返工
(最多 MAX_ROUNDS 轮防死锁),否则放行。全程 fail-open。

识别的写文件工具:原生 Write/Edit/MultiEdit **以及** universal-agent 的
``ChunkWrite`` MCP 工具(``mcp__*__ChunkWrite``)——后者是老 Write/Edit-only 收集
恒空的根因(与 #824 同源)。
"""
from __future__ import annotations

import json
import os
import re
import sys
from pathlib import Path

# The bundled ``citation_gate`` package lives at <PLUGIN_ROOT>/citation_gate.
PLUGIN_ROOT = os.environ.get("CLAUDE_PLUGIN_ROOT") or str(Path(__file__).resolve().parent.parent)
sys.path.insert(0, PLUGIN_ROOT)

HOME = os.path.expanduser("~")
# Per-user block-round counters stay under the user's home, not the plugin.
ROUNDS_DIR = os.path.join(HOME, ".claude", ".cache", "citation_gate", "rounds")
MAX_ROUNDS = 3
EXTS = {".tex", ".bib", ".md"}

# Native write tools whose ``file_path`` input names a file the agent authored.
_NATIVE_WRITE = {"Write", "Edit", "MultiEdit"}

_CITATION_RE = re.compile(r"\\bibitem|\\cite|\[\d+\]|@(inproceedings|article|misc|book)\b")


def _is_write_tool(name: str) -> bool:
    """True for native Write/Edit/MultiEdit and any MCP-namespaced ChunkWrite
    (e.g. ``mcp__universal-agent-chunk-write__ChunkWrite``)."""
    if name in _NATIVE_WRITE:
        return True
    return name.split("__")[-1] == "ChunkWrite"


def _collect_candidate_files(transcript_path: str, cwd: str) -> list[str]:
    seen: list[str] = []
    try:
        lines = Path(transcript_path).read_text(encoding="utf-8").splitlines()
    except OSError:
        return []
    for line in lines:
        line = line.strip()
        if not line:
            continue
        try:
            ev = json.loads(line)
        except ValueError:
            continue
        content = (ev.get("message") or {}).get("content")
        if not isinstance(content, list):
            continue
        for block in content:
            if not isinstance(block, dict) or block.get("type") != "tool_use":
                continue
            if not _is_write_tool(block.get("name") or ""):
                continue
            fp = (block.get("input") or {}).get("file_path")
            if not fp or not isinstance(fp, str):
                continue
            # ChunkWrite often gets a workspace-relative path; resolve against the
            # session cwd so the existence/extension checks below are correct.
            if not os.path.isabs(fp):
                fp = os.path.join(cwd or ".", fp)
            fp = os.path.normpath(fp)
            if (os.path.splitext(fp)[1].lower() in EXTS
                    and os.path.isfile(fp) and fp not in seen):
                seen.append(fp)
    return seen


def _has_citation_signal(fp: str) -> bool:
    try:
        text = Path(fp).read_text(encoding="utf-8")
    except OSError:
        return False
    if "<!-- citation-gate: skip -->" in text:
        return False
    return bool(_CITATION_RE.search(text))


def _rounds_file(session_id: str) -> str:
    return os.path.join(ROUNDS_DIR, f"{session_id}.json")


def _read_rounds(session_id: str) -> int:
    try:
        with open(_rounds_file(session_id), encoding="utf-8") as f:
            return int(json.load(f).get("n", 0))
    except (OSError, ValueError):
        return 0


def _bump_rounds(session_id: str) -> int:
    os.makedirs(ROUNDS_DIR, exist_ok=True)
    n = _read_rounds(session_id) + 1
    with open(_rounds_file(session_id), "w", encoding="utf-8") as f:
        json.dump({"n": n}, f)
    return n


def _reset_rounds(session_id: str) -> None:
    try:
        os.unlink(_rounds_file(session_id))
    except OSError:
        pass


def _run_verifier(files: list[str]) -> dict:
    """Import the bundled verifier in-process and return its report dict."""
    from citation_gate.verify import verify_files
    return verify_files(files).to_dict()


def _decide(report: dict, round_count: int, max_rounds: int) -> tuple[bool, str]:
    hard = report.get("hard_fail") or []
    soft = report.get("soft_warn") or []
    skip = report.get("skip") or []
    if not hard:
        msg = "引用校验通过。"
        if soft:
            msg += f" {len(soft)} 条未能核实,已标 [unverified],请人工确认。"
        if skip:
            msg += (f" 另有 {len(skip)} 条未核验(时间预算/网络限制),"
                    f"可调大 CITATION_GATE_BUDGET 或手动运行 python3 -m citation_gate 补检。")
        return False, msg
    detail = "\n".join(f"  [{h['index']}] {h['message']}" for h in hard)
    if round_count >= max_rounds:
        return False, (f"引用校验仍有 {len(hard)} 条疑似编造,但已达返工上限({max_rounds} 轮),放行。"
                       f"请人工(manual)复核:\n{detail}")
    return True, f"检测到 {len(hard)} 条疑似编造的引用,请逐条按权威记录修正后再交付:\n{detail}"


def _pass(sys_msg: str | None = None) -> None:
    print(json.dumps({"continue": True, "systemMessage": sys_msg} if sys_msg
                     else {"continue": True}, ensure_ascii=False))
    sys.exit(0)


def main() -> None:
    try:
        raw = sys.stdin.read()
        inp = json.loads(raw) if raw.strip() else {}
    except ValueError:
        inp = {}

    if os.environ.get("CITATION_GATE") == "off":
        _pass()
    session_id = inp.get("session_id") or "unknown"
    cwd = inp.get("cwd") or os.getcwd()

    files = [f for f in _collect_candidate_files(inp.get("transcript_path") or "", cwd)
             if _has_citation_signal(f)]
    if not files:
        _reset_rounds(session_id)
        _pass()

    try:
        report = _run_verifier(files)
    except Exception as e:  # fail-open: never break delivery on verifier error
        _pass(f"引用校验器未能运行({str(e)[:120]}),本次跳过。")

    round_count = _read_rounds(session_id)
    block, msg = _decide(report, round_count, MAX_ROUNDS)
    if block:
        _bump_rounds(session_id)
        print(json.dumps({"decision": "block", "reason": msg}, ensure_ascii=False))
        sys.exit(0)
    _reset_rounds(session_id)
    _pass(msg)


if __name__ == "__main__":
    main()
